src/mofaflex/_core/datasets/misc.py

```diff
@@ -51,11 +52,39 @@ def __init__(
     ):
         super().__init__()
 
-        self.covariates, self.covariates_names = data.get_covariates(obs_key, obsm_key)
-        self.covariates = {
-            group_name: np.nanmean(np.stack(tuple(group_covars.values()), axis=0), axis=0)
-            for group_name, group_covars in self.covariates.items()
-        }
+        covariates, self.covariates_names = data.get_covariates(obs_key, obsm_key)
+
+        # if data is categorical, get unique categories
+        categories = set()
+        for _, group_covars in covariates.items():
+            for _, view_covars in group_covars.items():
+                if view_covars.dtype == np.object_:
+                    non_nan_values = view_covars[~pd.isnull(view_covars)]
+                    categories.update(set(np.unique(non_nan_values)))
+                    have_nans = non_nan_values.size < view_covars.size
+
+        # map categories to floats
+        categories_mapping = {cat: float(i) if have_nans else i for i, cat in enumerate(sorted(categories))}
+
+        for group_name, group_covars in covariates.items():
+            for view_name, view_covars in group_covars.items():
+                if view_covars.dtype == np.object_:
+                    covariates[group_name][view_name] = np.vectorize(lambda cat: categories_mapping.get(cat, np.nan))(
+                        view_covars
+                    )
```

Comment:
Perhaps it would be faster to iterate over the entries of `categories_mapping` instead, something like
```suggestion
                    for k, v in categories_mapping.items():
                         new_arr[view_covars == k] = v
```
where `new_arr` is a pre-initialized array. `np.vectorize` has to drop back to Python for each element of the array, while the above is using Numpy's fast C backend.